azure/services/async/helper.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package async
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
+)
+
+// ProcessOngoingOperation is a helper function that will process an ongoing operation to check if it is done.
+// If it is not done, it will return a transient error.
+func ProcessOngoingOperation(ctx context.Context, scope FutureScope, client FutureHandler, resourceName string, serviceName string) error {
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future == nil {
+		scope.V(2).Info("no long running operation found", "service", serviceName, "resource", resourceName)
+		return nil
+	}
+	sdkFuture, err := converters.FutureToSDK(*future)
+	if err != nil {
+		// Reset the future data to avoid getting stuck in a bad loop.
```

Comment:
"bad loop"? Can you expand on what could cause such a loop? Is the future not in the format we expect and thus will continue to fail forever?

Reply:
In theory, this should never happen, but if for some reason the future that is already stored in Status isn't properly formatted and we don't reset it we would be stuck in an infinite loop trying to parse it and being unable to move on. I will expand the comment to make that more clear.

Reply:
added